rails/app/controllers/api/v1/jwt_controller.rb

```diff
@@ -145,39 +218,45 @@ def firebase
         class_hash = params[:class_hash]
       elsif params[:resource_link_id].present?
         # The resource_link_id param was already verified in the handle_initial_auth method
-        # The offering_id is not added to the claims because we don't want to restrict the
-        # teacher to just this one offering.
         offering = Portal::Offering.find(params[:resource_link_id])
         class_hash = offering.clazz.class_hash
       end
 
-      claims = {
-        # Firebase auth rules expect all the claims to be in a sub-object named "claims".
-        # All the new properties should go there. Other apps can still read them.
-        :claims => {
-          :platform_id => APP_CONFIG[:site_url],
-          :platform_user_id => user.id,
-          :user_type => "teacher",
-          :user_id => jwt_user_id(user),
-          :class_hash => class_hash
-        },
-        # Depreciated, used by some CC client apps. Do not add more data here, it's better to add that to claims
-        # object above, as then Firebase auth rules can read these properties too.
-        :domain => root_url,
-        :domain_uid => user.id,
-      }
+      sub_claims.merge!({
+        user_type: "teacher",
+        class_hash: class_hash,
+        # The offering_id is not added to the claims because we don't want to restrict the
+        # teacher to just this one offering.
+      })
+
+      # Depreciated, used by some CC client apps. Do not add more data here, you should
+      # add it to sub_claims so the Firebase auth rules can read the properties.
+      claims.merge!({
+        domain: root_url,
+        domain_uid: user.id,
+      })
     else
-      claims = {
-        # Firebase auth rules expect all the claims to be in a sub-object named "claims".
-        # All the new properties should go there. Other apps can still read them.
-        :claims => {
-          :platform_id => APP_CONFIG[:site_url],
-          :platform_user_id => user.id,
-          :user_type => "user",
-          :user_id => jwt_user_id(user)
-        }
-      }
-      # since the generic user case was added after domain and domain_uid where deprecated they are not set here
+
+      sub_claims.merge!({
+        user_type: "user",
+      })
+
+      resource_link_id = params[:resource_link_id]
+      if resource_link_id
+        # The resource_link_id param was already verified in the handle_initial_auth method
+
+        offering = Portal::Offering.find(resource_link_id)
+        class_hash = offering.clazz.class_hash
+        sub_claims.merge!({
+          class_hash: class_hash,
+          offering_id: offering.id,
+          # Any systems granting access based on this target_user_id claim should scope it to the
+          # class_hash and or offering_id so the main user doesn't gain access to all of the target user's
+          # data. Our policies don't always restrict access like this, but they should when possible.
+          target_user_id: params[:target_user_id].to_i
```

Comment:
I think we might end up with `target_user_id: 0` when params[:target_user_id] is `nil`. 
If there's a student with ID = 0, seems like a way to steal his data. 😉 
But seriously, maybe it's better to add this param to JWT only if it's explicitly provided in the request? I can imagine a developer inspecting JWT by hand in one of the client apps and wondering why there's some `target_user_id` defined at all. Well, I don't know if we can get to this point without `target_user_id` defined (auth failing somewhere earlier?), but just double checking.

Reply:
Yeah it is confusing. I couldn't find a better way to handle all the cases. If the resource_link_id is set and the current user is not a student or teacher of that resrouce_link, then handle_initial_auth will raise an exception if target_user_id is not set. 

I can make this a bit more clear by adding that to the conditional on line 245 and adding a comment about the handle_initial_auth